---
title: "Single-cell RNA-seq data analysis"
author: "Dresden-concept Genome Center, TU Dresden"
date: "June 2020"
geometry: margin=2cm
bibliography: "references.bib"
output:
   html_document:
    toc: true
    highlight: tango
    theme: paper
    code_folding: hide
---

```{r setup, warning=FALSE, message=FALSE}
# R Options
options(stringsAsFactors=FALSE, "citation_format"="pandoc")

# Required libraries
library(Seurat) # main
library(ggplot2) # plots
library(patchwork) # combination of plots
library(magrittr) # %>% operator
library(reticulate) # required for 'leiden' clustering
library(enrichR) # functional enrichment
library(future) # multicore support for Seurat
plan("multiprocess") # by default, this uses all available cores; use "workers=4" for example, to use 4 cores

# Other libraries we use
# Knit: knitr
# Data handling: dplyr, tidyr, purrr
# Tables: kableExtra
# Plots: ggsci, ggpubr
# IO: openxlsx
# Annotation: biomaRt
# DEG: mast
# Functional enrichment: enrichR

# Source plotting functions
source("R/functions_io.R")
source("R/functions_plotting.R")
source("R/functions_analysis.R")
source("R/functions_util.R")

# Knitr default options
knitr::opts_chunk$set(echo=TRUE, cache=FALSE, message=FALSE, warning=FALSE, fig.width=10)

# Potentially needed for clustering, umap, other python packages
use_python('/usr/bin/python')

# biomaRt mirror
mirror="asia"
```

# Dataset description
* 10X published dataset, downloaded from [here](https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_1k_v3) 
* Single Cell Gene Expression Dataset by Cell Ranger 3.0.0  
* 1k Peripheral blood mononuclear cells (PBMCs) from a healthy donor (v3 chemistry)  

## Project-specific parameters
This code chunk contains all parameters that are set specifically for your project. 
```{r project_parameters, class.source='fold-show'}
param = list()

# Project ID
param$project = "pbmc"

# Input data path in case Cell Ranger was run 
param$path_data = data.frame(name="pbmc",
                             type="10x",
                             path="test_datasets/10x_pbmc_1k_healthyDonor_v3Chemistry/counts/")
param$file_mapping_stats = "test_datasets/10x_pbmc_1k_healthyDonor_v3Chemistry/pbmc_1k_v3_metrics_summary.csv"

# Project-specific paths
param$path_out = "test_datasets/10x_pbmc_1k_healthyDonor_v3Chemistry/Seurat_results"
if (!file.exists(param$path_out)) dir.create(param$path_out, recursive=TRUE)

# Annotation via biomaRt
param$mart_dataset = "hsapiens_gene_ensembl"
param$annot_version = 98
param$annot_main = c(ensembl="ensembl_gene_id", symbol="external_gene_name", entrez="entrezgene_accession")
param$file_annot = NULL
if (is.null(param$file_annot)) {
  param$file_annot = file.path(param$path_out, paste0(param$mart_dataset, ".v", param$annot_version, ".annot.txt"))
}
param$mart_attributes = c(param$annot_main, 
                          c("chromosome_name", "start_position", "end_position", 
                            "percentage_gene_gc_content", "gene_biotype", "strand", "description"))

# Prefix of mitochondrial genes 
param$mt = "^MT-"

# The number of PCs to use; adjust this parameter based on JackStraw and Elbowplot 
param$pc_n = 10

# Whether or not to remove cell cycle effects
param$cc_remove = FALSE

# Should all cell cycle effects be removed, or only the difference between profilerating cells (G2M and S phase)?
# Read https://satijalab.org/seurat/v3.1/cell_cycle_vignette.html, for an explanation
param$cc_remove_all = FALSE

# Resolution of clusters; low values will lead to fewer clusters of cells 
param$cluster_resolution=0.5

# Thresholds to define differentially expressed genes 
param$padj = 0.05
param$log2fc = log2(1.5)

# Marker genes based on literature, translated to Ensembl IDs
# https://icb-scanpy-tutorials.readthedocs-hosted.com/en/latest/visualizing-marker-genes.html
# xlsx file, one list per column, first row as header and Ensembl IDs below
param$file_known_markers = "test_datasets/10x_pbmc_1k_healthyDonor_v3Chemistry/known_markers.xlsx"

# Enrichr databases of interest
param$enrichr_dbs = c("GO_Molecular_Function_2018", "GO_Biological_Process_2018", "GO_Cellular_Component_2018")

# Main colour(s) to use for plots
param$col = "palevioletred"

# The number of colours provided should match the number of datasets
if (nrow(param$path_data) > 1) {
  param$col_samples = ggsci::pal_npg()(nrow(param$path_data))
  names(param$col_samples) = param$path_data$name
} else { 
  param$col_samples = param$col
}
```

# Read data
## Read and print mapping statistics
We begin by printing mapping statistics that have been produced prior to this workflow. 
```{r mapping_stats}
if (!is.null(param$file_mapping_stats)) {
  mapping_stats = as.data.frame(t(read.delim(param$file_mapping_stats, sep=",", header=TRUE, check.names=FALSE)))
  colnames(mapping_stats) = "Value"
  knitr::kable(mapping_stats, align="l", caption="Mapping statistics") %>% 
    kableExtra::kable_styling(bootstrap_options=c("striped", "hover"))
 
} else { 
  message("Mapping statistics cannot be shown. No valid file provided.")
}
```

## Read gene annotation
If not provided already, we read gene annotation from Ensembl and write the resulting table to file. We generate several dictionaries to translate between Ensembl IDs, gene symbols, Entrez Ids, and Seurat rownames. 
```{r read_annotation}
# Read annotation from csv or from Ensembl and a tab separated txt will be created
if (file.exists(param$file_annot)) {
  annot_ensembl = read.delim(param$file_annot)
} else {
  annot_mart = biomaRt::useEnsembl("ensembl", dataset=param$mart_dataset, mirror=mirror, version=param$annot_version)
  annot_ensembl = biomaRt::getBM(mart=annot_mart, attributes=param$mart_attributes)
  write.table(annot_ensembl, file=param$file_annot, sep='\t', col.names=TRUE, row.names=FALSE, append=FALSE)
  message("Gene annotation file was created at: ", param$file_annot)
  # Note: depending on the attributes, there might be more than one row per gene
}

# Double-check if we got all required annotation, in case annotation file was read
check_annot_main = all(param$annot_main %in% colnames(annot_ensembl))
if (!check_annot_main) {
  stop("The annotation table misses at least one of the following columns: ", paste(param$annot_main, collapse=", "))
}

# Create translation tables
ensembl = param$annot_main["ensembl"]
symbol = param$annot_main["symbol"]
entrez = param$annot_main["entrez"]

# Ensembl id to gene symbol
ensembl_to_symbol = unique(annot_ensembl[, c(ensembl, symbol)])
ensembl_to_symbol = setNames(ensembl_to_symbol[, symbol], ensembl_to_symbol[, ensembl])

# Ensembl id to seurat-compatible unique rowname
ensembl_to_seurat_rowname = unique(annot_ensembl[, c(ensembl, symbol)])
ensembl_to_seurat_rowname[, symbol] = make.unique(gsub(pattern="_", replacement="-", x=ensembl_to_seurat_rowname[, symbol], fixed=TRUE))
ensembl_to_seurat_rowname = setNames(ensembl_to_seurat_rowname[, symbol], ensembl_to_seurat_rowname[, ensembl])

# Seurat-compatible unique rowname to ensembl id
seurat_rowname_to_ensembl = setNames(names(ensembl_to_seurat_rowname), ensembl_to_seurat_rowname)

# Gene symbol to ensembl id: named LIST to account for genes where one symbol translates to multiple Ensembl IDs
symbol_to_ensembl_df = unique(annot_ensembl[, c(ensembl, symbol)])
symbol_to_ensembl = split(symbol_to_ensembl_df[, ensembl], symbol_to_ensembl_df[, symbol])

# Gene symbol to (seurat compatible unique) gene symbol: named LIST to account for genes with multiple names
symbol_to_seurat_rowname = unique(annot_ensembl[, c(ensembl, symbol)])
symbol_to_seurat_rowname$seurat_rowname = ensembl_to_seurat_rowname[symbol_to_seurat_rowname[, ensembl]]
symbol_to_seurat_rowname = split(symbol_to_seurat_rowname$seurat_rowname, symbol_to_seurat_rowname[, symbol])

# Ensembl to entrez
ensembl_to_entrez = unique(annot_ensembl[, c(ensembl, entrez)])
ensembl_to_entrez[, entrez] = ifelse(nchar(ensembl_to_entrez[, entrez]) == 0, NA, ensembl_to_entrez[, entrez])
ensembl_to_entrez = split(ensembl_to_entrez[, entrez], ensembl_to_entrez[, ensembl])

# Seurat-compatible unique rowname to entrez
seurat_rowname_to_ensembl_match = match(seurat_rowname_to_ensembl, names(ensembl_to_entrez))
names(seurat_rowname_to_ensembl_match) = names(seurat_rowname_to_ensembl)
seurat_rowname_to_entrez = purrr::map(seurat_rowname_to_ensembl_match, function(i) {unname(ensembl_to_entrez[[i]])})
```

## Setup the Seurat object
We next read the scRNA-seq counts table(s) to initialise a Seurat object. 
```{r read_datasets}
# List of Seurat objects
sc = list()

datasets = param$path_data
for (i in seq(nrow(datasets))) {
  name = datasets[i,"name"]
  type = datasets[i,"type"]
  path = datasets[i,"path"]
  
  # Read 10X or smartseq2
  if (type == "10x") {
    
    # Read 10X data into a Seurat object
    sc[[name]] = Read10xDataset(path, project=name, row_name_column=1, convert_row_names=ensembl_to_seurat_rowname)
    
  } else if (type == "smartseq2") {
    
    # Read SmartSeq-2 data into a Seurat object
    # Note: drop unneccessary columns, e.g. columns 2-6 for featureCounts results
    sc[[name]] = ReadSmartseq2Dataset(path, project=name, row_name_column=1, convert_row_names=ensembl_to_seurat_rowname)
    
    # Parse plate information: sample name, plate number, plate row and plate column
    plate_information = parse_plate_information(Seurat::Cells(sc[[name]]), 
                                                pattern='^(\\S+)_(\\d+)_([A-Z])(\\d+)$', 
                                                sample_name_group=1, 
                                                plate_number_group=2, 
                                                row_name_group=3, 
                                                col_name_group=4)
    if (nrow(plate_information) > 0) {
      
      # Add info to metadata
      sc[[name]] = Seurat::AddMetaData(sc[[name]], plate_information)
      
      # Also set identity
      sc[[name]] = Seurat::SetIdent(sc[[name]], value="SampleName")
    }
  } 
}

sc
```

```{r integrate_multiple_samples, eval=length(sc)>1, include=length(sc)>1}
#  this chunk is for integrating multiple samples

# alternatives: 
# a) merge raw counts and then normalise (less preferred)
# b) do normalisation per sample, then integrate (better)
#   - integrate: just merge or advanced methods

# at the moment only simple merge based on raw counts is done
if (is.list(sc) & length(sc)>1) {
  sc = merge(x=sc[[1]], y=sc[2:length(sc)], project=param$project)
}
```

```{r integrate_single_sample, eval=length(sc)==1, include=length(sc)==1}
# this chunk is for single sample
if (is.list(sc) & length(sc)==1) {
  sc = sc[[1]]
}
```

# Pre-processing
## Quality control 
We start the analysis by removing unwanted cells from the dataset. Three commonly used QC metrics include the number of unique genes detected in each cell ("nFeature"), the total number of molecules detected in each cell ("nCount"), and the percentage of reads that map to the mitochrondrial genome ("percent.mt").
```{r qc}
# Calculate percentage of reads that map to mitochondrial genome 
sc = PercentageFeatureSet(sc, pattern=param$mt, col.name="percent.mt")

# Metadata 
knitr::kable(head(sc@meta.data, 5), align="l", caption="Meta-data, top 5 rows") %>% 
  kableExtra::kable_styling(bootstrap_options = c("striped", "hover")) %>% 
  kableExtra::scroll_box(width="100%")

```

```{r qc_filter}
# Filter cutoffs
cells_filter = c(nFeature_RNA=200, nCount_RNA=NA, percent.mt=20)
features_filter = c(min_counts=1, min_cells=3) # feature has to be found by at least one count in one cell 
```

We keep cells with more than `r cells_filter["nFeature_RNA"]` detected features, and cells with less than `r cells_filter["percent.mt"]`% of mitochondrial reads. We keep features detected with at least `r features_filter["min_counts"]` count(s) in at least `r features_filter["min_cells"]` cell(s). 
```{r qc_plot}
# Plot QC metrics
features = c("nFeature_RNA", "nCount_RNA", "percent.mt")
p = VlnPlot(sc, features=features, pt.size=0, col=param$col_samples, combine=FALSE) 
names(p) = features
for (i in features) {
  p[[i]] = PlotMystyle(p[[i]], title=i, legend_position="none") + xlab("")
  if (!is.na(cells_filter[i])) p[[i]] = p[[i]] + geom_hline(yintercept=cells_filter[i], lty=2, col="darkgrey") 
}
p = patchwork::wrap_plots(p, ncol = 3) + patchwork::plot_annotation("Distribution of feature values") 
p 

# Correlate QC metrics
p = list()
p[[1]] = Seurat::FeatureScatter(sc, feature1=features[2], feature2=features[1], cols=param$col_samples)
p[[2]] = Seurat::FeatureScatter(sc, feature1=features[2], feature2=features[3], cols=param$col_samples)
for (i in 1:length(p)) p[[i]] = PlotMystyle(p[[i]], legend_position="none")
p = patchwork::wrap_plots(p, ncol = 2) + patchwork::plot_annotation("Features plotted against each other") 
p

# Filter cells
sc = subset(sc, subset=nFeature_RNA>cells_filter["nFeature_RNA"] & percent.mt<cells_filter["percent.mt"])

# Filter features
features_count_sum = Matrix::rowSums(Seurat::GetAssayData(sc, slot="counts", assay="RNA") >= features_filter["min_counts"])
features_keep = which(features_count_sum > features_filter["min_cells"])
sc = subset(sc, features=features_keep)

sc
```

## Normalisation, cell cycle scoring, scaling and variable features {.tabset}
In this section, we subsequently run a series of Seurat functions:  
1. We start by running a __standard log normalisation__, where counts for each cell are divided by the total counts for that cell and multiplied by 10,000. This is then natural-log transformed.   
2. __Variable features__ are selected. For downstream analysis it is beneficial to focus on genes that exhibit high cell-to-cell variation, that is they are highly expressed in some cells and lowly expressed in others.   
3. To be able to compare normalised gene counts between genes, gene counts are further __scaled__ to have zero mean and unit variance (z-score). This way, genes are equally weighted for downstream analysis.  

In addition, we run:  
4. __SCTransform__, a new and more sophisticated normalisation method that replaces the previous functions (__normalisation, variable features and scaling__). Note that although results are kept for steps 1-3 ("RNA" assay), downstream analyses will default to resulting data of step 4 ("SCT" assay). 

If specified in the above parameter section, __cell cycle scores__ are assigned to each cell based on its normalised expression of G2/M and S phase markers, after basic normalisation (step 1). Cell cycle effects are then removed during scaling (step 4). Done for this report: `r param$cc_remove`. 

```{r norm_scale_variableFeatures, results="hide"}
# This function often crashes with multicore support, so stop for a moment
future::plan("sequential")

# Normalise data the original way
sc = Seurat::NormalizeData(sc, normalization.method = "LogNormalize", scale.factor = 10000)

# Back to multicore support
future::plan("multicore")

# Probably obsolete, not needed anywhere in this report
# Select features from normalised data (unaffected by scaling)
sc = Seurat::FindVariableFeatures(sc, selection.method = "vst", nfeatures = 3000)

# Scaling "RNA" normalised counts is required to run PCA, which again is required for the JackStraw plot
# Note that we do NOT remove cell cycle effects in "RNA" scaled data, since this is very slow and 
#   only used for the JackStraw plot
sc = Seurat::ScaleData(sc, features=rownames(sc))
```

```{r cc_genes_translate}
# Use biomart to translate human cell cycle genes to the species of interest
mart_human = biomaRt::useEnsembl("ensembl", dataset="hsapiens_gene_ensembl", mirror=mirror, version=param$ensembl_version) 
mart_myspecies = biomaRt::useEnsembl("ensembl", dataset=param$mart_dataset, mirror=mirror, version=param$ensembl_version) 

genes_s = biomaRt::getLDS(attributes=c("external_gene_name"), 
                          filters="external_gene_name", 
                          values=cc.genes.updated.2019$s.genes, 
                          mart=mart_human, 
                          attributesL=c("external_gene_name"), 
                          martL=mart_myspecies, 
                          uniqueRows=TRUE)
genes_g2m = biomaRt::getLDS(attributes=c("external_gene_name"), 
                            filters="external_gene_name", 
                            values=cc.genes.updated.2019$g2m.genes, 
                            mart=mart_human, 
                            attributesL=c("external_gene_name"), 
                            martL=mart_myspecies, 
                            uniqueRows=TRUE)

# Restrict genes to those present in the sc object
genes_s = genes_s[genes_s[,2] %in% rownames(sc), 2] 
genes_g2m = genes_g2m[genes_g2m[,2] %in% rownames(sc), 2]
```

```{r cc_scores, eval=param$cc_remove}
# If cell cycle effects should be removed, we first score cells 
# The effect is then removed in the following chunk 
if(param$cc_remove) {
# Determine cell cycle effect
sc = Seurat::CellCycleScoring(object=sc, s.features=genes_s, g2m.features=genes_g2m, set.ident=FALSE)

# Determine difference between G2M and S phase scores
sc$CC.Difference = sc$S.Score - sc$G2M.Score
}
```

```{r sctransform, results="hide"}
# Run sctransform
if (param$cc_remove) {

  # Regress out cell cycle effect
  if(param$cc_remove_all) {
    
    # Remove all signal associated to cell cycle
    sc = SCTransform(sc, vars.to.regress=c("S.Score", "G2M.Score"), min_cells=features_filter["min_cells"])
  } else { 
    
  # Remove only the difference between G2M and S phase
    sc = SCTransform(sc, vars.to.regress="CC.Difference", min_cells=features_filter["min_cells"])
  }
} else {
  sc = SCTransform(sc, min_cells=features_filter["min_cells"])
}

# "SCT" is now the default assay (active.assay)

# This is a new normalisation method that replaces previous Seurat functions 'NormalizeData', 'FindVariableFeatures', and 'ScaleData'. 
# vignette: https://satijalab.org/seurat/v3.0/sctransform_vignette.html
# paper: https://www.biorxiv.org/content/10.1101/576827v2
# Normalised data end up here: sc@assays$SCT@data

# Note: If I write Seurat::SCTransform, I get the following error message in the next chunk for
#   Seurat::VariableFeaturePlot(sc, cols=c("grey", param$col)) 
#   'vst' is Unable to find highly variable feature information for method 'vst'
```

The following plots provide a first impression of the genes that show the highest variance across the dataset, determined before scaling and potential removal of cell cycle effects. 
```{r plot_variable_features, fig.width=10}
# Show variable genes
top10 = head(Seurat::VariableFeatures(sc), 10)

# Plot variable features with and without labels
p1 = Seurat::VariableFeaturePlot(sc, cols=c("grey", param$col)) 
p1 = PlotMystyle(p1)
p2 = LabelPoints(plot=p1, points=top10, repel=TRUE, xnudge=0, ynudge=0) 
p = p1 + p2 & theme(legend.position="bottom")
p = p + patchwork::plot_annotation("Variable genes without (left) and with (right) labels.") + 
  patchwork::plot_layout(guides = "collect") 
p
```

```{r plot_norm}
n_cells_plot = min(200, length(Seurat::Cells(sc)))
cells_random = sample(1:length(Seurat::Cells(sc)), n_cells_plot)

raw = GetAssayData(sc, assay="RNA", slot="counts") %>% as.matrix()
raw = raw[, cells_random]

norm1 = GetAssayData(sc, assay="RNA", slot="data") %>% as.matrix()
norm1 = norm1[, cells_random]

norm2 = GetAssayData(sc, assay="SCT", slot="data") %>% as.matrix()
norm2 = norm2[, cells_random]
```

To better understand the efficiency of the applied normalisation procedures, we plot the relative log expression of genes in `r n_cells_plot` randomly selected cells before and after normalisation. This type of plot reveals unwanted variation in your data. The concept is taken from `r knitcitations::citet("10.1371/journal.pone.0191629")`. In brief, we remove variation between genes, leaving only variation between samples. If expression levels of most genes are similar in all cell types, sample heterogeneity is a sign of unwanted variation. 

For each gene, we calculate its median expression across all cells, and then calculate the deviation from this median for each cell. For each cell, we plot the median expression (black), the interquartile range (lightgrey), whiskers defined as 1.5 times the interquartile range (darkgrey), and outliers (`r param$col`). 

### Raw counts
```{r plot_norm_raw}
raw = log2(raw + 1)
p = PlotRLE(raw, col=param$col) + labs(title="log2(raw counts + 1)")
p
```

### Standard normalisation
```{r plot_norm_norm}
p = PlotRLE(norm1, col=param$col) + labs(title="Normalised data")
p 
```

### Advanced normalisation (SCT)
```{r plot_norm_sct}
p = PlotRLE(norm2, col=param$col) + labs(title="SCTransform'ed data")
p 
```

## Dimensionality reduction
A single-cell dataset of 20,000 genes and 5,000 cells has 20,000 dimensions. The biological manifold however can be described by far fewer dimensions than the number of genes. Dimension reduction methods aim to find these dimensions. There are two general purposes for dimension reduction methods: to summarise a dataset, and to visualise a dataset. 

We use Principal Component Analysis (PCA) to summarise a dataset, overcoming noise and reducing the data to its essential components. Each principal component (PC) represents a "metafeature" that combines information across a correlated gene set. Later, we use Uniform Manifold Approximation and Projection (UMAP) to visualise the dataset, placing similar cells together in 2D space, see below. 

To decide how many PCs to include in downstream analyses, we visualize cells and genes that define the PCA. 
```{r pca}
# Run PCA for "RNA" assay; this is required for the JackStraw plot
sc = Seurat::RunPCA(sc, features=Seurat::VariableFeatures(object=sc), assay="RNA")

# Run PCA for the default "SCT" assay
sc = Seurat::RunPCA(sc, features=Seurat::VariableFeatures(object=sc))

p = Seurat::VizDimLoadings(sc, dims=1:2, reduction="pca", col=param$col, combine=FALSE)
for (i in 1:length(p)) p[[i]] = PlotMystyle(p[[i]])
p =  patchwork::wrap_plots(p, ncol = 2) + patchwork::plot_annotation("Top gene loadings of the first two PCs") 
p

p = Seurat::DimPlot(sc, reduction="pca", cols=param$col_samples)
p = PlotMystyle(p, title="Cells arranged by the first two PCs", legend_position="bottom")
p

Seurat::DimHeatmap(sc, dims=1, cells=500, balanced=TRUE)
```

```{r pca_heatmaps, fig.height=20}
Seurat::DimHeatmap(sc, dims=1:20, cells=500, balanced=TRUE)
```

## Dimensionality of the dataset
We next need to decide how many PCs we want to use for analyses. The following two plots are designed to help us make an informed decision. 

The first plot is based on the "JackStraw" procedure: parts of the data is repeatedly randomly permuted and PCA is rerun, generating a "null distribution" of feature scores. Significant PCs are those with a strong enrichment of low p-value features. 

The second plot is an "Elbow plot": PCs are ranked based on the percentage of variance they explain. 

For your dataset, we decided to go for `r param$pc_n` PCs. 
```{r dimensionality, results="hide"}
# NOTE: This process can take a long time for big datasets, comment out for expediency. More
# approximate techniques such as those implemented in ElbowPlot() can be used to reduce
# computation time
sc = Seurat::JackStraw(sc, num.replicate=100, dims=20, assay="RNA")
sc = Seurat::ScoreJackStraw(sc, dims=1:20)
p = Seurat::JackStrawPlot(sc, dims=1:20)
p = PlotMystyle(p, title="Jack Straw plot", legend_position="bottom")
p

p = Seurat::ElbowPlot(sc, ndims=20)
p = PlotMystyle(p, title="Elbow plot")
p
```

# Downstream analysis
## Clustering
Seurat's clustering method first constructs a graph structure, where nodes are cells and edges are drawn between cells with similar gene expression patterns. Technically speaking, Seurat first constructs a K-nearest neighbor (KNN) graph based on Euclidean distance in PCA space, and refines edge weights between cells based on the shared overlap in their local neighborhoods (Jaccard similarity). To partition the graph into highly interconnected parts, cells are iteratively grouped together using the Leiden algorithm. 
```{r clustering}
# Note: I changed the seed in ./lib/python3.6/site-packages/leidenalg/functions.py to 11 for reproducibility
# The number of clusters can be optimized by tuning 'resolution' -> based on feedback from the client whether or not clusters make sense

# Choose the number of PCs to use for clustering
sc = Seurat::FindNeighbors(sc, dims=1:param$pc_n)

# Cluster using the Leiden algorithm
# Paper to Leiden algorithm: https://www.nature.com/articles/s41598-019-41695-z
# Seurat vignette suggests resolution parameter between 0.4-1.2 for datasets of about 3k cells
sc = Seurat::FindClusters(sc, resolution=param$cluster_resolution, algorithm=4)
```

## Visualisation with UMAP
We use a UMAP to visualise and explore a dataset. The goal is to place similar cells together in 2D space, and learn about the biology underlying the data. Cells are color-coded according to the graph-based clustering, and clusters typcially co-localise on the UMAP. 

Take care not to mis-read a UMAP:  

* Parameters influence the plot (we use defaults here)  
* Cluster sizes relative to each other mean nothing, since the method has a local notion of distance  
* Distances between clusters might not mean anything  
* You may need more than one plot  
  
For a nice read to intuitively understand UMAP, see `r knitcitations::citep("https://pair-code.github.io/understanding-umap/")`. 
```{r umap, message=FALSE, results="hide"}
sc = Seurat::RunUMAP(sc, dims=1:param$pc_n)

# Note that you can set `label = TRUE` or use the LabelClusters function to help label individual clusters
cluster_cells = table(sc@active.ident)
cluster_labels = paste0(levels(sc@active.ident)," (", cluster_cells[levels(sc@active.ident)],")")
p = Seurat::DimPlot(sc, reduction="umap", label=TRUE) + scale_colour_discrete("Cluster", labels=cluster_labels)
p = PlotMystyle(p, "UMAP, cells coloured by cluster identity", legend_position="bottom")
p
```

## Cell Cycle Effect {.tabset}
How much do gene expression profiles in your dataset reflect the cell cycle phases the single cells were in? We determine the effects of cell cycle heterogeneity by calculating a score for each cell based on its expression of G2M and S phase markers. Scoring is based on the strategy described in `r knitcitations::citet("10.1126/science.aad0501")`, and human gene names are translated using biomaRt. 
```{r cellCycleEffect}
# Determine cell cycle effect, this time based on default assay "SCT"
sc = CellCycleScoring(object=sc, s.features=genes_s, g2m.features=genes_g2m, set.ident=FALSE)

# Get a feeling for how many cells are affected
p1 = ggplot(sc@meta.data, aes(x=S.Score, y=G2M.Score, colour=Phase)) + 
  geom_point() + 
  scale_x_continuous("G1/S score") + 
  scale_y_continuous("G2/M score")
p1 = PlotMystyle(p1)
p2 = ggplot(sc@meta.data %>% 
              dplyr::group_by(seurat_clusters,Phase) %>% 
              dplyr::summarise(num_reads=length(Phase)), 
            aes(x=seurat_clusters, y=num_reads, fill=Phase)) + 
  geom_bar(stat="identity", position="fill") + 
  scale_x_discrete("Seurat clusters") + 
  scale_y_continuous("Fraction of cells")
p2 = PlotMystyle(p2)
p = p1 + p2 & theme(legend.position="bottom")
p = p + patchwork::plot_annotation(title="Cell cycle phases")
p
```

### UMAP coloured by cell cycle phases
```{r cellCycleEffect_umap_phases}
# UMAP with phases superimposed
p = Seurat::DimPlot(sc, group.by="Phase", pt.size=1)
PlotMystyle(p, title="UMAP, cells coloured by cell cycle phases", legend_title="Phase")
```

### UMAP coloured by S phase
```{r cellCycleEffect_umap_s}
p = Seurat::FeaturePlot(sc, features="S.Score", pt.size=1, min.cutoff="q1", max.cutoff="q99", cols=c("lightgrey", param$col))
p = PlotMystyle(p, title="UMAP, cells coloured by S phase")
p
```

### UMAP coloured by G2/M phase
```{r cellCycleEffect_umap_g2m}
p = Seurat::FeaturePlot(sc, features="G2M.Score", pt.size=1, min.cutoff="q1", max.cutoff="q99", cols=c("lightgrey", param$col))
p = PlotMystyle(p, title="UMAP, cells coloured by G2M phase")
p
```

## Feature plots QC  {.tabset}
Do cells in individual clusters have particularly high counts, detected genes or mitochondrial content?

### Number of counts
```{r featurePlot_QC_nCount}
p = Seurat::FeaturePlot(sc, features="nCount_RNA", cols=c("lightgrey", param$col))
p = PlotMystyle(p, title="nCount_RNA")
p
```

### Number of features
```{r featurePlot_QC_nFeature}
p = Seurat::FeaturePlot(sc, features="nFeature_RNA", cols=c("lightgrey", param$col))
p = PlotMystyle(p, title="nFeature_RNA")
p
```

### Percent mitochondrial reads
```{r featurePlot_QC_mt}
p = Seurat::FeaturePlot(sc, features="percent.mt", cols=c("lightgrey", param$col))
p = PlotMystyle(p, title="percent.mt")
p
```

## Feature plots for known marker genes {.tabset}
Do cells in individual clusters express provided known marker genes?
```{r featurePlot_knownMarkers_read}
# Read known marker genes
known_markers = openxlsx::read.xlsx(param$file_known_markers)
known_markers_list = lapply(colnames(known_markers), function(x) {
  y = ensembl_to_seurat_rowname[known_markers[,x]] %>% 
    na.exclude() %>% unique() %>% sort() 
  return(y)
})
names(known_markers_list) = colnames(known_markers)
known_markers_vect = unlist(known_markers_list) %>% unique() %>% sort()

fig_height_knownMarkers_list = 5*length(known_markers_list)
fig_height_knownMarkers_vect = max(5, length(known_markers_vect))
```

### Dot plot(s)
We show one dot plot per gene list provided. 
```{r featurePlot_knownMarkers_dotplot, fig.height=fig_height_knownMarkers_list}
p = list()
for (i in 1:length(known_markers_list)) {
  g = known_markers_list[[i]]
  g = g[length(g):1]
  p[[i]] = Seurat::DotPlot(sc, features=g, cols=c("lightgrey", param$col))
  p[[i]] = PlotMystyle(p[[i]], title=paste("Known marker genes:", names(known_markers_list)[i])) + 
    theme(axis.text.x = element_text(angle=90, hjust=1, vjust=.5))
}
patchwork::wrap_plots(p, ncol=1)
```

### Average feature plot(s)
We show one average feature plot per gene list provided, given that the list contains at least 10 genes. Positive scores suggest that genes are expressed more highly than expected. 
```{r featurePlot_knownMarkers_umap, fig.height=fig_height_knownMarkers_list}
idx_avg = sapply(1:length(known_markers_list), function(x) length(known_markers_list[[x]]) >= 10)
if (any(idx_avg)) {
  known_markers_avg = known_markers_list[idx_avg]
  sc = Seurat::AddModuleScore(sc, features=known_markers_avg, assay="SCT", ctrl=10, name="known_markers")
  idx_replace_names = grep("^known_markers[0-9]+$", colnames(sc@meta.data), perl=TRUE)
  colnames(sc@meta.data)[idx_replace_names] = names(known_markers_list)
  p = Seurat::FeaturePlot(sc, features=names(known_markers_list), cols=c("lightgrey", param$col), combine=FALSE)
  for (i in 1:length(known_markers_avg)) p[[i]] = PlotMystyle(p[[i]], 
                                                              title=paste("Known marker genes:", names(known_markers_list)[i]))
  patchwork::wrap_plots(p, ncol=1)
} else {
  print("This tab is used to plot an average for 10 or more genes. Your provided list(s) is (are) shorter than this, and hence average feature plots are skipped.")
}
```

### Individual feature plots
We show individual feature plots for all genes provided. 
```{r featurePlot_knownMarkers_all, fig.height=fig_height_knownMarkers_vect}
if (length(known_markers_vect) <= 30) {
  p = Seurat::FeaturePlot(sc, features=known_markers_vect, cols=c("lightgrey", param$col), combine=FALSE)
  for (i in 1:length(p)) p[[i]] = PlotMystyle(p[[i]])
  patchwork::wrap_plots(p, ncol=2)
} else { 
  print("This tab is used to plot up to 30 known marker genes. Your provided list is longer than this, and hence individual feature plots are skipped.")
}
```

## Differentially expressed genes
We next identify genes that are differentially expressed in one cluster compared to all other clusters. Additional gene annotation is added, and the resulting tables are written to file. 
```{r DEG}
# We load and unload the MAST R package in this chunk, as it overwrites Seurat functions
library(MAST)

# Find markers for every cluster compared to all remaining cells, report positive and negative ones
# min.pct = requires feature to be detected at this minimum percentage in either of the two groups of cells 
# logfc.threshold = requires a feature to be differentially expressed on average by some amount between the two groups
# only.pos = find only positive markers 

# Should cell cycle effects be ignored?
latent_vars=NULL
if(param$cc_remove) {
  if(param$cc_remove_all) {
    latent_vars = c("S.Score", "G2M.Score")
  } else {
    latent_vars = "CC.Difference"
  }
} 

# Review recommends using "MAST"; Mathias uses "LR"
# ALWAYS USE: assay="RNA" or assay="SCT"
# DONT USE: assay=integrated datasets; this data is normalised and contains only 2k genes
sc_markers = Seurat::FindAllMarkers(sc, assay="RNA", test.use="MAST",
                                    only.pos=FALSE, min.pct=0.25, logfc.threshold=0.25,
                                    latent.vars=latent_vars)

sc_markers_top2 = sc_markers %>% 
  dplyr::group_by(cluster) %>% 
  dplyr::top_n(n=2, wt=avg_logFC) %>% 
  as.data.frame()

# Show top 2 merkers per cluster
knitr::kable(sc_markers_top2, align="l", caption="Top 2 DEGs per cell cluster") %>% 
  kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))

# Add Ensembl annotation
sc_markers_ensembl = seurat_rowname_to_ensembl[sc_markers[,"gene"]]
sc_markers_annot = cbind(sc_markers, annot_ensembl[sc_markers_ensembl,])

# Output in Excel sheet
sc_markers_lst = lapply(levels(sc_markers_annot$cluster), function(x) {sc_markers_annot %>% dplyr::filter(cluster==x)})
names(sc_markers_lst) = paste0("cluster", levels(sc_markers$cluster))
openxlsx::write.xlsx(sc_markers_lst, file=paste0(param$path_out, "/markers.xlsx"))

# Filter markers based on p-value and fold-change 
sc_markers_filt = sc_markers %>% 
  dplyr::filter(p_val_adj <= param$padj) %>% 
  dplyr::filter((avg_logFC <= -param$log2fc) | (avg_logFC >= param$log2fc)) %>% 
  as.data.frame()
sc_markers_filt_down = sc_markers_filt %>% 
  dplyr::filter(avg_logFC <= -param$log2fc) %>% 
  as.data.frame()
sc_markers_filt_up = sc_markers_filt %>% 
  dplyr::filter(avg_logFC >= param$log2fc) %>% 
  as.data.frame()

# Number of DEGs per cluster
cluster_all = sort(unique(sc_markers[,"cluster"]))
sc_markers_filt_n = cbind(Cluster=cluster_all, 
                          Down=sapply(cluster_all, function(x) sum(sc_markers_filt_down$cluster == x)), 
                          Up=sapply(cluster_all, function(x) sum(sc_markers_filt_up$cluster == x))) %>% 
  as.data.frame() %>% 
  tidyr::pivot_longer(cols=c("Down", "Up"), 
                      names_to="Direction", 
                      values_to="n")
sc_markers_filt_n$Cluster = as.factor(sc_markers_filt_n$Cluster)

p = ggplot(sc_markers_filt_n, aes(x=Cluster, y=n, fill=Direction)) + geom_bar(stat="identity") 
p = PlotMystyle(p, 
                title=paste0("Number of DEGs per cell cluster\n(FC=", 2^param$log2fc, ", adj. p-value=", param$padj, ")"), 
                col=c("steelblue", "darkgoldenrod1"))
p

# Unload the MAST R package
detach("package:MAST", unload=TRUE)
```

## Visualisation of differentially expressed genes {.tabset}
The following plots are exemplary to how we can visualize differentially expressed genes using the "Seurat" R-package. The selected genes are the top differentially expressed genes for all clusters, respectively. 

```{r DEG_plot}
# Get top gene per cluster and plot
genes_example = sc_markers %>% 
  dplyr::group_by(cluster) %>% 
  dplyr::top_n(n=1, wt=avg_logFC) %>% 
  dplyr::pull(gene) %>% 
  unique()

fig_height_deg = max(5, length(genes_example))
```

### Feature plots
```{r DEG_plot_umap, fig.height=fig_height_deg}
# Shows gene expression on the UMAP
p = Seurat::FeaturePlot(sc, features=genes_example, cols=c("lightgrey", param$col), combine=FALSE)
names(p) = genes_example
for (i in names(p)) p[[i]] = PlotMystyle(p[[i]], title=i)
p = patchwork::wrap_plots(p, ncol=2) + 
  patchwork::plot_annotation(title="UMAP, cells coloured by normalised gene expression data")
p
```

### Ridge plots (raw)
```{r DEG_plot_ridge_raw, fig.height=fig_height_deg}
# Ridge plot of raw gene expression counts
p = Seurat::RidgePlot(sc, features=genes_example, slot="counts", combine=FALSE)
names(p) = genes_example
for (i in names(p)) p[[i]] = PlotMystyle(p[[i]], title=i, legend_title="Cell identity")
p = patchwork::wrap_plots(p, ncol=2) + 
  patchwork::plot_annotation(title="Ridge plot of raw gene expression counts") + 
  patchwork::plot_layout(guides = "collect") &
  theme(legend.position="bottom")
p
```

### Ridge plots (normalised)
```{r DEG_plot_ridge_norm, fig.height=fig_height_deg}
# Ridge plot of normalised gene expression data
p = Seurat::RidgePlot(sc, features=genes_example, combine=FALSE)
names(p) = genes_example
for (i in names(p)) p[[i]] = PlotMystyle(p[[i]], title=i, legend_title="Cell identity")
p = patchwork::wrap_plots(p, ncol=2) + 
  patchwork::plot_annotation(title="Ridge plot of normalised gene expression data") + 
  patchwork::plot_layout(guides = "collect") &
  theme(legend.position="bottom")
p
```

### Dot plot
```{r DEG_plot_dot}
# Visualises how feature expression changes across different clusters
p = Seurat::DotPlot(sc, features=genes_example[length(genes_example):1], cols=c("lightgrey", param$col))
p = PlotMystyle(p, title="Dot plot of normalised gene expression data") + 
  theme(axis.text.x = element_text(angle=90, hjust=1, vjust=.5))
p
```

### Heatmap
```{r DEG_heatmap, fig.height=20}
# Heatmap of top differentially expressed genes
top = sc_markers %>% 
  dplyr::group_by(cluster) %>% 
  dplyr::top_n(n=10, wt=avg_logFC)
p = Seurat::DoHeatmap(sc, features=top$gene) + NoLegend() + ggtitle("Heatmap of scaled gene expression data")
p
```

## Functional enrichment analysis
To gain first insights into potential functions of cells in a cluster, we test for over-representation of functional terms amongst up- and down-regulated genes of each cluster. Over-represented terms are written to file.  

We first translate gene symbols of up- and down-regulated genes per cluster into Entrez gene symbols, and then use the "enrichR" R-package to access the "Enrichr" website `r knitcitations::citep("https://amp.pharm.mssm.edu/Enrichr/")`. You can choose to test functional enrichment from a wide range of databases:
```{r enrichr_databases}
dbs_all = enrichR::listEnrichrDbs()
knitr::kable(dbs_all, align="l", caption="Enrichr databases") %>% 
  kableExtra::kable_styling(bootstrap_options=c("striped", "hover")) %>% 
  kableExtra::scroll_box(width="100%", height="300px")
```

```{r functional_enrichment, results="hide"}
# DEGs up and down per cluster
cluster_all = sort(unique(sc_markers[,"cluster"]))
genesets_up = lapply(cluster_all, function(x) {
  tmp = sc_markers_filt_up %>% 
    dplyr::filter(cluster==x) %>% 
    dplyr::pull(gene)
  # Pick the first matching Entrez symbol
  tmp = sapply(tmp, function(x) seurat_rowname_to_entrez[[x]][1]) %>% 
    na.exclude() %>% unique()
  return(tmp)
})
genesets_down = lapply(cluster_all, function(x) {
  tmp = sc_markers_filt_down %>% 
    dplyr::filter(cluster==x) %>% 
    dplyr::pull(gene)
  # Pick the first matching Entrez symbol
  tmp = sapply(tmp, function(x) seurat_rowname_to_entrez[[x]][1]) %>% 
    na.exclude() %>% unique()
  return(tmp)
})
names(genesets_up) = paste0("DEG_up_cluster_", cluster_all)
names(genesets_down) = paste0("DEG_down_cluster_", cluster_all)
genesets = c(genesets_up, genesets_down)
  
# Loop through gene lists
enriched = list()
for (i in 1:length(genesets)) {
  if (length(genesets[[i]]) >= 3) {
    message("Geneset ", names(genesets)[i])
    enriched[[i]] = enrichR::enrichr(genesets[[i]], databases=param$enrichr_dbs)
  } else { 
    message("Geneset ", names(genesets)[i], " has less than 3 genes; skip enrichr")
    enriched[[i]] = NA
  }
}
names(enriched) = names(genesets)

# Write enrichment results to file
enriched_top = matrix(NA, nrow=0, ncol=6)
colnames(enriched_top) = c("GeneSet", "Database", "Term", "Overlap", "Adjusted_pval", "Genes")
for (i in 1:length(enriched)) { 
  if (!is.null(enriched[[i]])) { 
    openxlsx::write.xlsx(enriched[[i]], file=paste0(param$path_out, "/Functions_", names(enriched)[i], ".xlsx"))
  }
}
```

The following table contains the top enriched term per geneset and database. 
```{r functional_enrichment_results}
for (i in 1:length(enriched)) { 
  if ((!is.null(enriched[[i]])) & (!is.na(enriched[[i]]))) { 

    # Remember top term per geneset
    for(j in 1:length(enriched[[i]])) {
        enriched_top = rbind(enriched_top, 
                             c(names(enriched)[i], 
                             names(enriched[[i]])[j], 
                             enriched[[i]][[j]][1, c("Term", "Overlap", "Adjusted.P.value", "Genes")]))
    }
  }
}

# Print table of top terms per gene set
knitr::kable(enriched_top, align="l", caption="Top enriched term per geneset") %>% 
  kableExtra::kable_styling(bootstrap_options = c("striped", "hover")) %>% 
  kableExtra::scroll_box(width="100%", height="700px")
```

# Loupe Cell Browser integration
We export the UMAP 2D visualisation, metadata such as the cell clusters, and lists of differentially expressed genes, so you can open and work with these in the Loupe Cell Browser.  
```{r loupe_integration}
# Export UMAP coordinates
loupe_umap = as.data.frame(sc@reductions$umap@cell.embeddings)
loupe_umap = cbind(Barcode=rownames(loupe_umap), loupe_umap)
colnames(loupe_umap) = c("Barcode", "UMAP-1", "UMAP-2")
write.table(loupe_umap, file=paste0(param$path_out, "/Seurat2Loupe_umap.csv"), col.names=TRUE, row.names=FALSE, quote=FALSE, sep=",")

# Export categorical metadata
loupe_meta = as.data.frame(sc@meta.data)
idx_keep = sapply(1:ncol(loupe_meta), function(x) !is.numeric(loupe_meta[,x]))
loupe_meta = cbind(Barcode=rownames(loupe_meta), loupe_meta[, idx_keep])
write.table(x=loupe_meta, file=paste0(param$path_out, "/Seurat2Loupe_metadata.csv"), col.names=TRUE, row.names=FALSE, quote=FALSE, sep=",")

# Export gene sets
loupe_genesets = data.frame(List=paste0("DEG_up_cluster_", sc_markers_filt_up[,"cluster"]), 
                            Name=sc_markers_filt_up[,"gene"], 
                            Ensembl=seurat_rowname_to_ensembl[sc_markers_filt_up[,"gene"]])
loupe_genesets = rbind(loupe_genesets, 
                       data.frame(List=paste0("DEG_down_cluster_", sc_markers_filt_down[,"cluster"]), 
                                  Name=sc_markers_filt_down[,"gene"], 
                                  Ensembl=seurat_rowname_to_ensembl[sc_markers_filt_down[,"gene"]]))

genesets_to_export = list(genes_cc_s_phase=genes_s, genes_cc_g2m_phase=genes_g2m)
for (i in names(genesets_to_export)) {
  tmp_genes = genesets_to_export[[i]]
  tmp_genes = tmp_genes[tmp_genes %in% names(symbol_to_ensembl)]
  loupe_genesets = rbind(loupe_genesets,
                         data.frame(List=i,
                                    Name=tmp_genes,
                                    Ensembl=seurat_rowname_to_ensembl[tmp_genes]))
}

write.table(loupe_genesets, file=paste0(param$path_out, "/Seurat2Loupe_genesets.csv"), col.names=TRUE, row.names=FALSE, quote=FALSE, sep=",")
```

# Output files
All files generated with this report are written into the provided output folder `r param$path_out`: 

* Annotation files  
  + `r basename(param$file_annot)`: Table that contains several identifiers and annotation (columns) per gene (rows) (optional)   
* Differentially expressed genes  
  + markers_xlsx: Excel file with one tab per cell cluster  
* Loupe Cell Browser files  
  + Seurat2Loupe_umap.csv: Seurat UMAP 2D visualisation   
    Import to Loupe through "Import Projection"   
  + Seurat2Loupe_metadata.csv: Seurat categorial meta data including clusters and cell cycle phases   
    Import to Loupe through "Import Categories"   
  + Seurat2Loupe_genesets.csv: Seurat differentially expressed genes   
    Import to Loupe through "Import Lists"   
* Functional enrichment of differentially expressed genes per cell cluster  
  + Functions_DEG_down_cluster_1.xlsx, Functions_DEG_up_cluster_1.xlsx, ...: Excel files with one tab per database  
  
```{r save, eval=FALSE}
save.image(paste0(param$path_out, "/Seurat_scAnalysis.RData"))
```

# Software versions
This report was generated using the [scrnaseq](https://github.com/ktrns/scrnaseq) GitHub repository. Software versions were collected at run time. 
```{r versions}
out = scrnaseq_session_info()

knitr::kable(out, align="l") %>% 
  kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))
```

# References
```{r bib}
knitcitations::write.bibtex(file="references.bib")
```
